packages/upstream/ssl.0.5.9+1/files/0001-REQ-403-provide-custom-verification-callback.patch

```diff
@@ -0,0 +1,89 @@
+From 2f277eb39be139c8857df7a79bce039e83119a11 Mon Sep 17 00:00:00 2001
+From: Ben Anson <ben.anson@citrix.com>
+Date: Mon, 19 Apr 2021 15:03:49 +0100
+Subject: [PATCH] REQ-403 provide custom verification callback
+
+Intended usage from ocaml:
+
+```
+let ctx = (* my SSL_CTX *) in
+Ssl.load_verify_locations ctx "bundle.pem" "" ;
+Ssl.set_verify ctx [ Ssl.Verify_peer ] (Some Ssl.Citrix.exact_match_cb) ;
+
+(* when the bundle changes, we should refresh the trusted certs *)
+Ssl.load_verify_locations ctx "bundle.pem" "";
+```
+
+Signed-off-by: Ben Anson <ben.anson@citrix.com>
+---
+ src/ssl.ml      |  6 ++++++
+ src/ssl.mli     |  6 ++++++
+ src/ssl_stubs.c | 27 +++++++++++++++++++++++++++
+ 3 files changed, 39 insertions(+)
+
+diff --git a/src/ssl.ml b/src/ssl.ml
+index 4ecd57d..df75c43 100644
+--- a/src/ssl.ml
++++ b/src/ssl.ml
+@@ -331,3 +331,9 @@ let input_int ssl =
+     i := (!i lsl 8) + int_of_char (Bytes.get tmp 2);
+     i := (!i lsl 8) + int_of_char (Bytes.get tmp 3);
+     !i
++
++module Citrix = struct
++  external get_exact_match_cb : unit -> verify_callback = "ocaml_ssl_get_exact_match_cb"
++
++  let exact_match_cb = get_exact_match_cb ()
++end
+diff --git a/src/ssl.mli b/src/ssl.mli
+index eb62baa..3a68be9 100644
+--- a/src/ssl.mli
++++ b/src/ssl.mli
+@@ -473,3 +473,9 @@ val input_int : socket -> int
+ 
+ (** Write an integer on an SSL socket. *)
+ val output_int : socket -> int -> unit
++
++module Citrix : sig
++  val exact_match_cb : verify_callback
++  (** This callback can be used to check that a peer's certificate _exactly_ matches
++   *  a trusted certificate. The trusted certificates can be set using [load_verify_locations] *)
++end
+diff --git a/src/ssl_stubs.c b/src/ssl_stubs.c
+index 3d431b8..6b1598e 100644
+--- a/src/ssl_stubs.c
++++ b/src/ssl_stubs.c
+@@ -1664,3 +1664,30 @@ err:
+   if (bio != NULL) BIO_free(bio);
+   return(ret);
+ }
++
++static int exact_match_cb(int ok, X509_STORE_CTX *ctx) {
```

Comment:
the signature can be seen here: https://man.archlinux.org/man/SSL_CTX_set_cert_verify_callback.3ssl.en

Reply:
> What is the type verify_callback?

On the ocaml side, it's actually a pointer to a C function

> Is this the right signature?

The signature has to match the signature of the existing callback in `ssl_stubs.c`

```
static int client_verify_callback(int ok, X509_STORE_CTX *ctx)
```

`ok` is somewhat like an error code (except that 1 is VERIFY_OK). It might be prudent to assert that the value passed in is 1 (i.e. that any prechecks were successful), but I will need to test that

Reply:
This suggests that `ok` is chained and should not be overwritten:

```
if (!ok) return ok;
```

Reply:
I tested this, and I think the prechecks do a hostname check, so unfortunately we must ignore the original value of `ok`

Reply:
the example given in the documentation assumes the value of `ok` is 1, if the application-defined function finds an invalid state it changes its value to 0. (in the documentation is called `preverify_ok`)